README.md
<!-- Badges -->
[![CRAN
Version](https://img.shields.io/cran/v/miselect?style=flat-square&color=blue&label=CRAN)](https://cran.r-project.org/package=miselect)
[![GitHub
Release](https://img.shields.io/github/v/release/umich-cphds/miselect?include_prereleases&label=Github&style=flat-square&color=blue)](https://github.com/umich-cphds/miselect)
[![Travis
CI](https://img.shields.io/travis/umich-cphds/miselect?style=flat-square)](https://travis-ci.org/umich-cphds/miselect)

Variable Selection for Multiply Imputed Data
============================================

Penalized regression methods, such as lasso and elastic net, are used in
many biomedical applications when simultaneous regression coefficient
estimation and variable selection is desired. However, missing data
complicates the implementation of these methods, particularly when
missingness is handled using multiple imputation. Applying a variable
selection algorithm on each imputed dataset will likely lead to
different sets of selected predictors, making it difficult to ascertain
a final active set without resorting to ad hoc combination rules.
'miselect' presents Stacked Adaptive Elastic Net (saenet) and Grouped
Adaptive LASSO (galasso) for continuous and binary outcomes. They, by
construction, force selection of the same variables across multiply
imputed data. 'miselect' also provides cross validated variants of these
methods.

Installation
------------

`miselect` can installed from Github via

    # install.packages("devtools")
    devtools::install_github("umich-cphds/miselect", build_opts = c())

The Github version may contain bug fixes not yet present on CRAN, so if
you are experiencing issues, you may want to try the Github version of
the package.

Example
-------

Here is how to use cross validated `saenet`. A nice feature of `saenet`
is that you can cross validate over `alpha` without having to use
`foldid`.

    library(miselect)
    library(mice)
    #> 
    #> Attaching package: 'mice'
    #> The following objects are masked from 'package:base':
    #> 
    #>     cbind, rbind

    set.seed(48109)
    # Using the mice defaults for sake of example only.
    mids <- mice(miselect.df, m = 5, printFlag = FALSE)
    dfs <- lapply(1:5, function(i) complete(mids, action = i))

    # Generate list of imputed design matrices and imputed responses
    x <- list()
    y <- list()
    for (i in 1:5) {
        x[[i]] <- as.matrix(dfs[[i]][, paste0("X", 1:20)])
        y[[i]] <- dfs[[i]]$Y
    }

    # Calculate observational weights
    weights  <- 1 - rowMeans(is.na(miselect.df))
    pf       <- rep(1, 20)
    adWeight <- rep(1, 20)
    alpha    <- c(.5 , 1)

    # Since 'Y' is a binary variable, we use 'family = "binomial"'
    fit <- cv.saenet(x, y, pf, adWeight, weights, family = "binomial",
                     alpha = alpha)

    # By default 'coef' returns the betas for (lambda.min , alpha.min)
    coef(fit)
    #> (Intercept)          X1          X2          X3          X4          X5 
    #>  0.07500679  1.48805119  0.57645656  0.00000000  1.79085524  0.05315903 
    #>          X6          X7          X8          X9         X10         X11 
    #>  0.00000000  1.78466568  0.00000000 -0.01937788  0.00000000  1.29382032 
    #>         X12         X13         X14         X15         X16         X17 
    #>  0.00000000  0.00000000  0.00000000 -0.14915055  0.00000000  0.28699265 
    #>         X18         X19         X20 
    #>  0.00000000 -0.23440273  0.00000000

You can supply different values of `lambda` and `alpha`. Here we use the
`lambda` and `alpha` selected by the one standard error rule

    coef(fit, lambda = fit$lambda.1se, alpha = fit$alpha.1se)
    #> (Intercept)          X1          X2          X3          X4          X5 
    #>   0.1180812   0.0000000   0.0000000   0.0000000   0.0000000   0.0000000 
    #>          X6          X7          X8          X9         X10         X11 
    #>   0.0000000   0.0000000   0.0000000   0.0000000   0.0000000   0.0000000 
    #>         X12         X13         X14         X15         X16         X17 
    #>   0.0000000   0.0000000   0.0000000   0.0000000   0.0000000   0.0000000 
    #>         X18         X19         X20 
    #>   0.0000000   0.0000000   0.0000000

Bugs
----

If you encounter a bug, please open an issue on the
[Issues](https://github.com/umich-cphds/miselect/issues) tab on Github
or send us an email.

Contact
-------

For questions or feedback, please email Jiacong Du at
<jiacong@umich.edu> or Alexander Rix <alexrix@umich.edu>.

References
----------

Variable selection with multiply-imputed datasets: choosing between
stacked and grouped methods. Jiacong Du, Jonathan Boss, Peisong Han,
Lauren J Beesley, Stephen A Goutman, Stuart Batterman, Eva L Feldman,
and Bhramar Mukherjee. 2020. arXiv:2003.07398


NEWS.md
# miselect v0.9.1

## Minor Changes
* Rewrote parts of `saenet` to increase computation speed.
* Refactor `saenet` and `galasso` to increase code readability for beta update.
* Update tests to include more complicated test cases.

## Bug fixes
* Fix bug when user specifies penalty factor (pf) = 0 for a variable.
* Fix bug in `cv.saenet` when calculating `alpha.1se` and `lambda.1se`.

# miselect v0.9
* Initial CRAN release

# miselect v0.7
* Beta version of miselect


R/cv.saenet.R
#' Cross Validated Multiple Imputation Stacked Adaptive Elastic Net
#'
#' Does k-fold cross-validation for \code{saenet}, and returns optimal values
#' for lambda and alpha.
#'
#' \code{cv.saenet} works by stacking the multiply imputed data into a single
#' matrix and running a weighted adaptive elastic net on it. Simulations suggest
#' that the "stacked" objective function approaches tend to be more
#' computationally efficient and have better estimation and selection
#' properties.
#'
#' Due to stacking, the automatically generated \code{lambda} sequence
#' \code{cv.saenet} generates may end up underestimating \code{lambda.max}, and
#' thus the degrees of freedom  may be nonzero at the first lambda value.
#' @param x A length \code{m} list of \code{n * p} numeric matrices. No matrix
#'     should contain an intercept, or any missing values
#' @param y A length \code{m} list of length \code{n} numeric response vectors.
#'     No vector should contain missing values
#' @param pf Penalty factor of length \code{p}. Can be used to differentially
#'     penalize certain variables. 0 indicates to not penalize the covariate
#' @param adWeight Numeric vector of length p representing the adaptive weights
#'     for the L1 penalty
#' @param weights Numeric vector of length n containing the proportion observed
#'     (non-missing) for each row in the un-imputed data.
#' @param family The type of response. "gaussian" implies a continuous response
#'     and "binomial" implies a binary response. Default is "gaussian".
#' @param alpha Elastic net parameter. Can be a vector to cross validate over.
#'     Default is 1
#' @param nlambda Length of automatically generated 'lambda' sequence. If
#'     lambda' is non NULL, 'nlambda' is ignored. Default is 100
#' @param lambda.min.ratio Ratio that determines the minimum value of 'lambda'
#'     when automatically generating a 'lambda' sequence. If 'lambda' is not
#'     NULL, 'lambda.min.ratio' is ignored. Default is 1e-3
#' @param lambda Optional numeric vector of lambdas to fit. If NULL,
#'    \code{galasso} will automatically generate a lambda sequence based off
#'    of \code{nlambda} and code{lambda.min.ratio}. Default is NULL
#' @param nfolds Number of foldid to use for cross validation. Default is 5,
#'     minimum is 3
#' @param foldid an optional length \code{n} vector of values between 1 and
#     ‘nfold’ identifying what fold each observation is in. Default is NULL and
#'     \code{cv.galasso} will automatically generate folds
#' @param maxit Maximum number of iterations to run. Default is 1000
#' @param eps Tolerance for convergence. Default is 1e-5
#' @return An object of type "cv.saenet" with 9 elements:
#' \describe{
#' \item{call}{The call that generated the output.}
#' \item{lambda}{Sequence of lambdas fit.}
#' \item{cvm}{Average cross validation error for each lambda and alpha. For
#'            family = "gaussian", 'cvm' corresponds to mean squared error,
#'            and for binomial 'cvm' corresponds to deviance.}
#' \item{cvse}{Standard error of 'cvm'.}
#' \item{saenet.fit}{A 'saenet' object fit to the full data.}
#' \item{lambda.min}{The lambda value for the model with the minimum cross
#'                   validation error.}
#' \item{lambda.1se}{The lambda value for the  sparsest model within one
#'                   standard error of the minimum cross validation error.}
#' \item{alpha.min}{The alpha value for the model with the minimum cross
#'                   validation error.}
#' \item{alpha.1se}{The alpha value for the  sparsest model within one
#'                   standard error of the minimum cross validation error.}

#' \item{df}{The number of nonzero coefficients for each value of lambda and alpha.}
#' }
#' @references
#' Variable selection with multiply-imputed datasets: choosing between stacked
#' and grouped methods. Jiacong Du, Jonathan Boss, Peisong Han, Lauren J Beesley,
#' Stephen A Goutman, Stuart Batterman, Eva L Feldman, and Bhramar Mukherjee. 2020.
#' arXiv:2003.07398
#' @examples
#' \donttest{
#' library(miselect)
#' library(mice)
#'
#' set.seed(48109)
#'
#' # Using the mice defaults for sake of example only.
#' mids <- mice(miselect.df, m = 5, printFlag = FALSE)
#' dfs <- lapply(1:5, function(i) complete(mids, action = i))
#'
#' # Generate list of imputed design matrices and imputed responses
#' x <- list()
#' y <- list()
#' for (i in 1:5) {
#'     x[[i]] <- as.matrix(dfs[[i]][, paste0("X", 1:20)])
#'     y[[i]] <- dfs[[i]]$Y
#' }
#'
#' # Calculate observational weights
#' weights  <- 1 - rowMeans(is.na(miselect.df))
#' pf       <- rep(1, 20)
#' adWeight <- rep(1, 20)
#'
#' # Since 'Y' is a binary variable, we use 'family = "binomial"'
#' fit <- cv.saenet(x, y, pf, adWeight, weights, family = "binomial")
#'
#' # By default 'coef' returns the betas for (lambda.min , alpha.min)
#' coef(fit)
#' }
#'
#' # You can also cross validate over alpha
#' \donttest{
#' fit <- cv.saenet(x, y, pf, adWeight, weights, family = "binomial",
#'                  alpha = c(.5, 1))
#' # Get selected variables from the 1 standard error rule
#' coef(fit, lambda = fit$lambda.1se, alpha = fit$alpha.1se)
#'
#' }
#' @export
cv.saenet <- function(x, y, pf, adWeight, weights, family = c("gaussian", "binomial"),
                      alpha = 1, nlambda = 100, lambda.min.ratio =
                      ifelse(isTRUE(all.equal(adWeight, rep(1, p))), 1e-3, 1e-6),
                      lambda = NULL, nfolds = 5, foldid = NULL, maxit = 1000,
                      eps = 1e-5)
{
    call <- match.call()

    if (!is.list(x))
        stop("'x' should be a list of numeric matrices.")
    if (any(sapply(x, function(.x) !is.matrix(.x) || !is.numeric(.x))))
        stop("Every 'x' should be a numeric matrix.")

    dim <- dim(x[[1]])
    n <- dim[1]
    p <- dim[2]
    m <- length(x)

    if (!is.numeric(nfolds) || length(nfolds) > 1)
        stop("'nfolds' should a be single number.")

    if (!is.null(foldid))
        if (!is.numeric(foldid) || length(foldid) != length(y[[1]]))
            stop("'nfolds' should a be single number.")

    fit <- saenet(x, y, pf, adWeight, weights, family, alpha, nlambda,
                  lambda.min.ratio, lambda, maxit, eps)

    X <- do.call("rbind", x)
    Y <- do.call("c", y)

    weights <- rep(weights / m , m)

    if (!is.null(foldid)) {
        if (!is.numeric(foldid) || !is.vector(foldid) || length(foldid) != n)
            stop("'foldid' must be length n numeric vector.")
        nfolds <- max(foldid)
    } else {

        r     <- n %% nfolds
        q     <- (n - r) / nfolds
        foldid <- c(rep(seq(nfolds), q), seq(r))
        foldid <- sample(foldid, n)
        foldid <- rep(foldid, m)
    }
    if (nfolds < 3)
        stop("'nfolds' must be bigger than 3.")

    lambda  <- fit$lambda
    nlambda <- length(lambda)
    X.scaled <- scale(X, scale = apply(X, 2, function(.X) stats::sd(.X) * sqrt(m)))

    cvm  <- array(0, c(nlambda, length(alpha), nfolds))
    cvse <- matrix(nlambda, length(alpha))
    for (j in seq(nfolds)) {
        Y.train <- Y[foldid != j]
        X.train <- subset_scaled_matrix(X.scaled, foldid != j)
        w.train <- weights[foldid != j]

        X.test  <- X[foldid == j, , drop = F]
        Y.test  <- Y[foldid == j]
        w.test <- weights[foldid == j]

        cv.fit <- switch(match.arg(family),
            gaussian = fit.saenet.gaussian(X.train, Y.train, n, p, m, w.train,
                                           nlambda, lambda, alpha, pf, adWeight,
                                           maxit, eps),
            binomial = fit.saenet.binomial(X.train, Y.train, n, p, m, w.train,
                                           nlambda, lambda, alpha, pf, adWeight,
                                           maxit, eps)
        )

        cvm[,, j] <- cv.saenet.err(cv.fit, X.test, Y.test, w.test, m)
    }

    cvse <- apply(cvm, c(1, 2), stats::sd) / sqrt(nfolds)
    cvm  <- apply(cvm, c(1, 2), mean)

    i.min    <- which.min(apply(cvm, 1, min))
    j.min    <- which.min(apply(cvm, 2, min))

    lambda.min <- fit$lambda[i.min]
    alpha.min <- fit$alpha[j.min]

    k <- abs(cvm - cvm[i.min, j.min]) < cvse[i.min, j.min]

    k <- (fit$df + 1) * ifelse(k, 1, Inf)

    i.1se    <- which.min(apply(k, 1, min))
    j.1se    <- which.min(apply(k, 2, min))

    lambda.1se <- fit$lambda[i.1se]
    alpha.1se  <- alpha[j.1se]

    structure(list(call = call, lambda = fit$lambda, alpha = alpha, cvm = cvm,
                   cvse = cvse, saenet.fit = fit, lambda.min = lambda.min,
                   alpha.min = alpha.min, lambda.1se = lambda.1se, alpha.1se =
                   alpha.1se, df = fit$df), class = "cv.saenet")
}


cv.saenet.err <- function(cv.fit, X.test, Y.test, w.test, m)
{
    nalpha <- length(cv.fit$alpha)
    nlambda <- length(cv.fit$lambda)
    cvm <- matrix(0, nlambda, nalpha)

    for (j in seq(nlambda)) {
        for (i in seq(nalpha)) {
            beta <- cv.fit$beta[j, i,]
            beta0 <- beta[1]
            beta  <- beta[-1]
            if ("saenet.gaussian" %in% class(cv.fit)) {
                mse <- m * mean((Y.test - X.test %*% beta - beta0) ^ 2 * w.test)
                cvm[j, i] <- mse
            }
            else {
                eta <- X.test %*% beta + beta0
                dev <-  w.test * (Y.test * eta - log(1 + exp(eta)))
                cvm[j, i] <- -2 * m * mean(dev)
            }
        }
    }
    cvm
}


#' Print cv.saenet Objects
#'
#' \code{print.cv.saenet} print the fit and returns it invisibly.
#' @param x An object of type "cv.saenet" to print
#' @param ... Further arguments passed to or from other methods
#' @export
print.cv.saenet <- function(x, ...)
{
    nl <- length(x$lambda)
    na <- length(x$alpha)

    cvm <- x$cvm
    dimnames(cvm) <- list(paste0("l.", seq(nl)), paste0("a.", seq(na)))
    cat("'cv.saenet' fit:\n")
    print(x$call)
    cat("Average cross validation error for each (lambda, alpha)\n")
    print(cvm)
    cat("(lambda, alpha) min:\n")
    cat("(", x$lambda.min, ", ", x$alpha.min, ")\n", sep = "")
    cat("(lambda, alpha) 1 SE:\n")
    cat("(", x$lambda.1se, ", ", x$alpha.1se, ")\n", sep = "")
    invisible(x)
}
